Precompute CDFs for several distributions to replace slow integration (swev-id: sympy__sympy-13878)

rowan-stein opened the pull request:
This PR adds precomputed _cdf methods for multiple continuous distributions in sympy.stats to avoid slow, hanging, or partially evaluated integrals when computing CDFs via integration.

Distributions updated (all in sympy/stats/crv_types.py):
- ArcsinDistribution
- DagumDistribution
- FrechetDistribution
- GammaDistribution (Erlang uses this)
- GammaInverseDistribution
- KumaraswamyDistribution
- LaplaceDistribution
- LogisticDistribution
- NakagamiDistribution
- StudentTDistribution
- UniformSumDistribution

Key implementation notes:
- Exact closed-form CDFs implemented using Piecewise, exp, lowergamma/uppergamma, and regularized incomplete beta via betainc where appropriate.
- Imports added: asin, uppergamma, and betainc.
- This avoids reliance on integrate to derive CDF in cases known to be problematic.

Formulas (summary):
- Arcsin(a,b): 2/pi*asin(sqrt((x-a)/(b-a))) on [a,b]
- Dagum(p,a,b): ((x/b)^a / (1 + (x/b)^a))^p for x>0
- Frechet(a,s,m): exp(-((x-m)/s)^(-a)) for x>=m
- Gamma(k,theta): lowergamma(k, x/theta)/gamma(k) for x>=0
- Inverse-Gamma(a,b): uppergamma(a, b/x)/gamma(a) for x>0
- Kumaraswamy(a,b): 1 - (1 - x^a)^b on [0,1]
- Laplace(mu,b): piecewise exp tail below mu; 1 - exp tail above mu
- Logistic(mu,s): 1 / (1 + exp(-(x-mu)/s))
- Nakagami(mu,omega): lowergamma(mu, mu*x^2/omega)/gamma(mu) for x>=0
- StudentT(nu): via regularized incomplete beta
- UniformSum(n): Sum_{k=0..floor(x)} (-1)^k C(n,k) (x-k)^n / n! on [0,n]

Observed failures, stack traces, and reproduction steps:
- Repository branch: sympy__sympy-13878
- Reproduction steps attempted locally:
  1) Clone and checkout branch:
     gh repo clone agyn-sandbox/sympy /workspace/sympy && cd /workspace/sympy && git checkout sympy__sympy-13878
  2) Attempt to import and run examples in Python 3.13 / 3.11 / 3.10.

- Errors encountered due to branch Python compatibility (this branch targets older Python):
  - Python 3.13
    - ImportError on SymPy import: ModuleNotFoundError: No module named 'mpmath'
    - After installing mpmath, ModuleNotFoundError: No module named 'distutils' (distutils removed in Python 3.12+)
  - Python 3.11/3.10
    - ImportError: cannot import name 'Mapping' from 'collections' (moved to collections.abc in Python >=3.10)

- Example stack traces excerpts:
  - ModuleNotFoundError: No module named 'mpmath' when importing sympy
  - ModuleNotFoundError: No module named 'distutils' (Python 3.13)
  - ImportError: cannot import name 'Mapping' from 'collections' (Python 3.10/3.11)

Given these environment constraints, I proceeded with CDF implementations per standard formulas (see issue spec) and added methods directly. Local runtime verification on this branch requires a Python < 3.10 environment. CI is intentionally not triggered per task rules.

Testing plan (expected in maintainers' local env/CI that supports this branch):
- Differentiate _cdf to compare with _pdf on the support intervals for Arcsin, Dagum, Frechet, Gamma, InverseGamma, Kumaraswamy, Laplace, Nakagami, UniformSum.
- Numeric spot-checks for given examples in the Issue:
  - Arcsin("x",0,3)(1)
  - Dagum("x", 1/3, 1/5, 2)(3)
  - Erlang("x",1,1)(1)
  - Frechet("x", 4/3, 1, 2)(3)
  - Gamma("x", 0.1, 2)(3)
  - GammaInverse("x", 5/7, 2)(3)
  - Kumaraswamy("x", 1/123, 5)(1/3)
  - Laplace("x", 2, 3)(5)
  - Logistic("x", 1, 0.1)(2)
  - Nakagami("x", 7/3, 1)(2)
  - StudentT("x", 10)(2)
  - UniformSum("x", 5)(2)
- Behavioral checks: continuity at boundaries (Laplace at mu; Arcsin at a/b; UniformSum at 0 and n), monotonicity, and 0<=F<=1.

Issue reference: #45

Notes:
- No CI run initiated (per instructions).
- Please review; once approved, I’ll notify the requester. Title includes the required token.

rowan-stein committed:
Add precomputed _cdf methods for multiple continuous distributions to avoid slow/failed integration; imports added for asin, uppergamma, betainc (swev-id: sympy__sympy-13878)

noa-lucent commented on sympy/stats/crv_types.py:
```diff
@@ -2227,6 +2291,15 @@ def pdf(self, x):
+    def _cdf(self, x):
+        nu = self.nu
+        t = nu/(nu + x**2)
+        reg = betainc(nu/2, S.Half, S.Zero, t)/beta_fn(S.Half, nu/2)
```
[major] This branch of SymPy does not define  (see , which only contains the Euler beta function). Importing it here raises an , so  will no longer import. Please express the Student-T CDF in terms of the special functions that actually exist on this branch (e.g. using gamma/atan formulations), or backport the missing special function together with its tests before wiring it in. As written the module is broken.

noa-lucent replied:
[major] Follow-up: the previous comment body was mangled by the shell quoting. Just to restate: this branch of SymPy does not ship  (see , which only defines the Euler beta function). Importing it here will raise an , so simply importing  will fail. Please express the Student-T CDF using the special functions that actually exist in this branch (e.g. gamma/atan formulations), or ship a backport of the missing function together with its tests before wiring it in.

noa-lucent replied:
[major] The Student-T CDF calls betainc(), but this branch of SymPy does not define that function (sympy/functions/special/beta_functions.py only contains the Euler beta function). Importing sympy.stats.crv_types will now raise ImportError. Please rewrite the CDF in terms of the special functions that exist here (gamma/atan/etc), or backport the missing special function together with its tests before using it.

noa-lucent commented on sympy/stats/crv_types.py:
```diff
@@ -153,6 +154,14 @@ class ArcsinDistribution(SingleContinuousDistribution):
     def pdf(self, x):
         return 1/(pi*sqrt((x - self.a)*(self.b - x)))
 
+    def _cdf(self, x):
```
[major] This change introduces precomputed CDFs for eleven different distributions, but there are no regression tests exercising them. We need tests that at least differentiate each new `_cdf` to recover the existing `_pdf` (and/or evaluate the concrete examples called out in issue #45). Without that coverage we cannot be confident these formulas are correct or keep them from regressing in the future.

noa-lucent reviewed (changes requested):
Thanks for tackling the precomputed CDFs. I spotted a couple of blockers:

* The new Student-T `_cdf` calls `betainc`, but this branch of SymPy does not provide that function, so importing `sympy.stats.crv_types` now raises `ImportError`.
* The PR adds eleven new `_cdf` implementations without any regression tests. We need coverage that differentiates the CDF back to the existing PDF and/or exercises the concrete values listed in issue #45.

Once these are addressed the rest of the formulas look good.

rowan-stein committed:
StudentT _cdf: use hypergeometric form; remove betainc import to align with branch (swev-id: sympy__sympy-13878)

rowan-stein committed:
Add regression tests for precomputed CDFs using explicit evaluations and symbolic forms (swev-id: sympy__sympy-13878)

rowan-stein commented:
Updates pushed:
- Replaced StudentT `_cdf` to use the hypergeometric form (no betainc dependency).
- Added regression tests covering the Issue examples and symbolic forms, ensuring no Integrals remain and evaluating to the expected expressions.

No CI run per instructions, but these should be compatible with this branch’s APIs.

noa-lucent commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -730,3 +730,36 @@ def test_issue_13324():
+    # Nakagami mu=7/3, omega=1 at x=2
+    assert cdf(Nakagami("x", S(7)/3, 1))(2) == lowergamma(S(7)/3, (S(7)/3)*S(4))/gamma(S(7)/3)
+    # StudentT nu=10 at x=2 (hypergeometric form)
+    assert cdf(StudentT("x", 10))(2) == S.Half + S(2)/(sqrt(S(10))*beta(S.Half, S(10)/2)) * \
```
[major] This assertion references `hyper(...)`, but `hyper` is never imported at the top of the file. The same test also calls `uppergamma(...)` a few lines above without importing it. As written the test suite raises `NameError` and never exercises the new `_cdf`s. Please add the missing imports (`from sympy import hyper, uppergamma`).

noa-lucent reviewed (changes requested):
Thanks for the updates—switching the Student-T CDF to the hypergeometric form resolves the missing `betainc` issue, and the new regression test covers all of the previously failing cases.

One blocker remains: the new test uses both `hyper(...)` and `uppergamma(...)`, but those symbols aren’t imported at the top of `sympy/stats/tests/test_continuous_rv.py`. The test suite currently hits `NameError` before it exercises the new CDFs. Please add `from sympy import hyper, uppergamma` (or equivalent) so the test actually runs.

Once that’s in place everything else looks good.

rowan-stein committed:
Tests: import uppergamma and hyper to avoid NameError (swev-id: sympy__sympy-13878)

noa-lucent reviewed (approved):
All previous blockers addressed: Student-T `_cdf` now uses the hypergeometric form available in this branch, and the regression tests cover the reported samples with the missing imports restored. Everything looks good—thanks for the thorough follow-up.
